MatrixSDK/ThreePidAdd/MX3PidAddManager.m

```diff
@@ -0,0 +1,706 @@
+    {
+        operation = [mxSession.identityService submit3PIDValidationToken:token
+                                                                  medium:threePidAddSession.medium
+                                                            clientSecret:threePidAddSession.clientSecret
+                                                                     sid:threePidAddSession.sid
+                                                                 success:success
+                                                                 failure:failure];
+    }
+    else
+    {
+        NSLog(@"[MX3PidAddManager] submitValidationToken: ERROR: Failed to submit validation token for 3PID: %@, identity service is not set", threePidAddSession);
+
+        NSError *error = [NSError errorWithDomain:MX3PidAddManagerErrorDomain
+                                             code:MX3PidAddManagerErrorDomainIdentityServerRequired
+                                         userInfo:nil];
+        failure(error);
+    }
+
+    return operation;
+}
+
+
+#pragma mark - Bind to Identity Server -
+
+// https://gist.github.com/jryans/839a09bf0c5a70e2f36ed990d50ed928#3b-changing-the-bind-status-of-a-3pid-post-msc2290
+- (void)startIdentityServer3PidSession:(MX3PidAddSession*)threePidAddSession
+                     success:(void (^)(BOOL needValidation))success
+                     failure:(void (^)(NSError * _Nonnull))failure
+{
+    if (!mxSession.identityService)
+    {
+        NSError *error = [NSError errorWithDomain:MX3PidAddManagerErrorDomain
+                                             code:MX3PidAddManagerErrorDomainIdentityServerRequired
+                                         userInfo:nil];
+        failure(error);
+        return;
+    }
+
+    MXWeakify(self);
+    threePidAddSession.httpOperation = [self doesServerSupportSeparateAddAndBind:^(bool doesServerSupportSeparateAddAndBind) {
+        MXStrongifyAndReturnIfNil(self);
+
+        MXHTTPOperation *operation;
+        if (doesServerSupportSeparateAddAndBind)
+        {
+            if (threePidAddSession.bind)
+            {
+                operation = [self startIdentityServer3PidSessionWithNewHomeserver:threePidAddSession success:^{
+                    threePidAddSession.httpOperation = nil;
+                    success(YES);
+                } failure:^(NSError *error) {
+                    threePidAddSession.httpOperation = nil;
+                    failure(error);
+                }];
+            }
+            else
+            {
+                // No need of 3Pid validation in this configuration
+                operation = [self->mxSession.matrixRestClient unbind3PidWithAddress:threePidAddSession.address medium:threePidAddSession.medium success:^{
+
+                    threePidAddSession.httpOperation = nil;
+
+                    NSLog(@"[MX3PidAddManager] tryFinaliseIdentityServer3PidSessionWithNewHomeserver: DONE: threePid: %@", threePidAddSession);
```

Comment:
`startIdentityServer3PidSession` instead of `tryFinaliseIdentityServer3PidSessionWithNewHomeserver`.